imports/client/components/FirehosePage.tsx

```diff
@@ -0,0 +1,267 @@
+      hunt: huntId,
+    });
+    const ready = chatMessagesHandle.ready();
+    const chatMessages = ready ?
+      ChatMessages.find({ hunt: huntId }, { sort: { timestamp: 1 } }).fetch() : [];
+    return {
+      ready,
+      chatMessages,
+    };
+  }, [huntId]);
+
+  const messagesPaneRef = useRef<HTMLDivElement>(null);
+  const searchBarRef = useRef<HTMLInputElement>(null);
+
+  const maybeStealCtrlF = useCallback((e: KeyboardEvent) => {
+    if (e.ctrlKey && e.key === 'f') {
+      e.preventDefault();
+      const node = searchBarRef.current;
+      if (node) {
+        node.focus();
+      }
+    }
+  }, []);
+
+  useEffect(() => {
+    window.addEventListener('keydown', maybeStealCtrlF);
+    return () => {
+      window.removeEventListener('keydown', maybeStealCtrlF);
+    };
+  }, [maybeStealCtrlF]);
+
+  const setSearchString = useCallback((val: string) => {
+    const u = new URLSearchParams(props.location.search);
+    if (val) {
+      u.set('q', val);
+    } else {
+      u.delete('q');
+    }
+    props.history.replace({
+      pathname: props.location.pathname,
+      search: u.toString(),
+    });
+  }, [props.history, props.location]);
+
+  const onSearchStringChange: FormControlProps['onChange'] = useCallback((e) => {
+    setSearchString(e.currentTarget.value);
+  }, [setSearchString]);
+
+  const getSearchString = useCallback((): string => {
+    const u = new URLSearchParams(props.location.search);
+    const s = u.get('q');
+    return s || '';
+  }, [props.location.search]);
+
+  const clearSearch = useCallback(() => {
+    setSearchString('');
+  }, [setSearchString]);
+
+  const compileMatcher = useCallback((searchKeys: string[]): (c: ChatMessageType) => boolean => {
+    // Given a list a search keys, compileMatcher returns a function that,
+    // given a chat message, returns true if all search keys match that message in
+    // some way, and false if any of the search keys cannot be found.
+    const lowerSearchKeys = searchKeys.map((key) => key.toLowerCase());
+    return (chatMessage) => {
+      return lowerSearchKeys.every((key) => {
+        return chatMessage.text.toLowerCase().indexOf(key) !== -1;
+      });
+    };
+  }, []);
+
+  const filteredChats = useCallback((chatMessages: ChatMessageType[]) => {
+    const searchKeys = getSearchString().split(' ');
+    let interestingChatMessages;
+
+    if (searchKeys.length === 1 && searchKeys[0] === '') {
+      interestingChatMessages = chatMessages;
+    } else {
+      const searchKeysWithEmptyKeysRemoved = searchKeys.filter((key) => { return key.length > 0; });
+      const isInteresting = compileMatcher(searchKeysWithEmptyKeysRemoved);
+      interestingChatMessages = chatMessages.filter(isInteresting);
+    }
+
+    return interestingChatMessages;
+  }, [getSearchString, compileMatcher]);
+
+  const [shouldScrollBottom, setShouldScrollBottom] = useState<boolean>(true);
+
+  const saveShouldScroll = useCallback(() => {
```

Comment:
ShouldScrollBottom is not consistently set when a filter change causes the length of the list to increase and overflow, forcing the bottom off the list. (The preferred behavior in this case, since the bottom was completely visible prior, may actually be to preserve the state and advance the scroll.)

Reply:
Yeah, my code here was not right.  I like the point you make: if the end of the scrollback is visible, we should ensure it stays visible.  So:

* if content to display is shorter than the pane, `shouldScrollBottom` should get set to `true`.
* if content to display is larger than the pane, if the bottom of the pane is visible, `shouldScrollBottom` should get set to `true`.
* if content to display is larger than the pane, and the user scrolls off the bottom by more than a tiny bit, we should set `shouldScrollBottom` to `false`
* if the content to be displayed changes, or the space available to the content changes (due to e.g. window resize), apply the above rules again

Reply:
I can confirm this is fixed. The nature of scroll preservation when the filter is changing and the bottom isn't visible is not transparent, but it seems to behave reasonably, even returning to the same position if the filter is shortened, then restored. I'm not even sure specifically what the desired behavior would be, so I think it's good.